xamarin/Xamarin.Forms
# [WIP] Unify the image handling

### Description of Change ###

This is now a big PR that changes everything to use `ImageSource` instead of strings, files or other image-y types.

**NOTE: This is a set of breaking changes as the property types have changed!** 
__NOTE: The changes to the public API can be reverted temporarily as it is going down a level. As a result, all the code will still be valid.__

### Issues Resolved ### 

 - Fixes the inconsistency with image source types.
 - Fixes #3207 
- Fixes #4689

### Known Issues ###
https://github.com/xamarin/Xamarin.Forms/issues/6004
https://github.com/xamarin/Xamarin.Forms/issues/6010
regresses #4597 on API 19

### API Changes ###

Added:

```
class ImageSource {
    // Returns true if this image source is actually not going to return anything
    // It is not going to determine if the result will actually have a valid image, but 
    // rather it will indicate if there is no file path, or if the stream is null, or if 
    / the uri is empty
    public virtual bool IsEmpty { get; }
}
```

Changed the property types to `ImageSource` for the following properties:

```csharp
Button.Image // from FileImageSource
MenuItem.Icon // from FileImageSource
NavigationPage.TitleIcon // from FileImageSource (this also had the static getter/setter method for the attached property)
Page.BackgroundImage // from string
Page.Icon // from FileImageSource
Slider.ThumbImage // from FileImageSource
```

As a result of things changing, I also changed the name of `ImageElement.FileImageProperty` to `ImageProperty`. This should not affect anyone as it was internal.

### Platforms Affected ### 

All the platforms. Everything.

### Behavioral/Visual Changes ###

None.

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

Enable the `SetMainPage(new ImageSourcesGallery());` in the gallery app and make sure all the buttons do what they are supposed to do.

### PR Checklist ###
